### PR TITLE
Adding Test in frontend components 

### PR DESCRIPTION
**Pull request by Claudio-code**
Issue https://github.com/flathub/website/issues/1184

### Flows to be tested

- [x] Login redirect component
- [x]  Logout Button in header

### CODE REVIEW

File: frontend/__tests__/components/layout/Header.test.tsx [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+      <>
+        <UserContext.Provider value={userState}>
+          <UserDispatchContext.Provider value={dispatchMock}>
+            <Header />
+          </UserDispatchContext.Provider>
+        </UserContext.Provider>
+      </>,
+    )
+
+    await waitFor(() => {
```
Comment (razzeee) [inline frontend/__tests__/components/layout/Header.test.tsx:58]: We could also check for the avatar icon with user name here

File: frontend/__tests__/components/layout/Header.test.tsx [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+      fireEvent.click(getByText("open-user-menu"))
+      fireEvent.click(getByText("log-out"))
+    })
+
+    expect(dispatchMock).toHaveBeenCalledWith({ type: "loading" })
+    expect(dispatchMock).toHaveBeenCalledWith({ type: "logout" })
```
Comment (razzeee) [inline frontend/__tests__/components/layout/Header.test.tsx:64]: Check if we have the login button now?

File: frontend/__tests__/components/login/ProviderLink.test.tsx [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+  method: "github",
+  name: "GitHub",
+}
+const githubProps: Props = {
+  inACard: true,
+  provider: githubProvider,
+}
+const translationMock = jest.fn()
+
+jest.mock("react-i18next", () => ({
```
Comment (razzeee) [inline frontend/__tests__/components/login/ProviderLink.test.tsx:22]: We probably want to mock this globally

File: frontend/__tests__/components/login/ProviderLink.test.tsx [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+      i18n: {
+        changeLanguage: () => new Promise(() => {}),
+      },
+    }
+  },
+}))
+
+const useStateSpy = jest.spyOn(React, "useState")
+const setStateMock = jest.fn()
+// @ts-ignore
```
Comment (razzeee) [inline frontend/__tests__/components/login/ProviderLink.test.tsx:35]: Why do we need this?

File: frontend/__tests__/components/login/ProviderLink.test.tsx [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+        provider={githubProps.provider}
+        inACard={githubProps.inACard}
+      />,
+    )
+
+    await waitFor(() => {
+      fireEvent.click(getByRole("button"))
+    })
+    expect(setStateMock.mock.calls).toEqual([[true], [false]])
+    expect(toast.error).toHaveBeenCalled()
```
Comment (razzeee) [inline frontend/__tests__/components/login/ProviderLink.test.tsx:128]: These expectations would also pass on the previous test, so that's probably an anti pattern. Not sure if in the test or in the components behavior.

File: frontend/__tests__/components/login/ProviderLink.test.tsx [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+  })
+
+  it("redirect login test if clicked is true not call api", async () => {
+    global.fetch = jest.fn() as jest.Mock
+    // @ts-ignore
```
Comment (razzeee) [inline frontend/__tests__/components/login/ProviderLink.test.tsx:133]: Why is this needed?

**Comment by razzeee** [comment]
Great work, other then the mentioned comments, I think this misses one crucial thing. It needs to be added to the github pipeline. Maybe as a new pipeline step, so that we can validate these pass on every merge.

**Comment by Claudio-code** [comment]
Cool i make adjustments during the week

- [ ] We could also check for the avatar icon with user name here
- [ ] Check the error test on the component
- [ ] Add mock global to react-i18next
- [ ] Remove all @ts-ignore
- [ ] In logout test add check if have login button
- [ ] Add tests in github pipeline

File: frontend/__tests__/components/login/ProviderLink.test.tsx (hunk 6 above)
Comment (Claudio-code) [reply frontend/__tests__/components/login/ProviderLink.test.tsx]: was removed

File: frontend/__tests__/components/login/ProviderLink.test.tsx (hunk 4 above)
Comment (Claudio-code) [reply frontend/__tests__/components/login/ProviderLink.test.tsx]: was removed

File: frontend/__tests__/components/login/ProviderLink.test.tsx (hunk 3 above)
Comment (Claudio-code) [reply frontend/__tests__/components/login/ProviderLink.test.tsx]: changed to global mock

File: frontend/__tests__/components/login/ProviderLink.test.tsx (hunk 5 above)
Comment (Claudio-code) [reply frontend/__tests__/components/login/ProviderLink.test.tsx:117]: I agree, I'll just leave the validation of the toast text, which is the only difference between the error scenarios

File: frontend/__tests__/components/layout/Header.test.tsx (hunk 2 above)
Comment (Claudio-code) [reply frontend/__tests__/components/layout/Header.test.tsx:69]: cool added it test too

File: frontend/__tests__/components/layout/Header.test.tsx (hunk 1 above)
Comment (Claudio-code) [reply frontend/__tests__/components/layout/Header.test.tsx:61]: cool added it test

**Comment by Claudio-code** [comment]
@razzeee I made the adjustments if you can review again

File: .github/workflows/frontend_ci.yml [hunk 7]
```diff
@@ -43,6 +43,9 @@ jobs:
       - name: Lint
         run: yarn lint
 
+      - name: Unitary Tests
```
Comment (razzeee) [inline .github/workflows/frontend_ci.yml:46]: I've never seen it called this either `Unit tests` or better just `Frontend tests`

File: frontend/__tests__/components/layout/Header.test.tsx [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+      <>
+        <UserContext.Provider value={userState}>
+          <UserDispatchContext.Provider value={dispatchMock}>
+            <Header />
+          </UserDispatchContext.Provider>
+        </UserContext.Provider>
+      </>,
+    )
+
+    const userIconBeforeLogout = container.querySelector("img.rounded-full")
```
Comment (razzeee) [inline frontend/__tests__/components/layout/Header.test.tsx:55]: I would much rather see us using testing library here to also ensure accessability

https://testing-library.com/docs/queries/about#priority

File: .github/workflows/frontend_ci.yml (hunk 7 above)
Comment (Claudio-code) [reply .github/workflows/frontend_ci.yml]: Changed stage name

File: frontend/__tests__/components/layout/Header.test.tsx (hunk 8 above)
Comment (Claudio-code) [reply frontend/__tests__/components/layout/Header.test.tsx]: changed to getByRole

**Comment by razzeee** [comment]
Hrm, seems like it's not running on our CI

**Comment by Claudio-code** [comment]
there was a conflict in the yarn.lock file

**Comment by razzeee** [comment]
I guess we're running into https://github.com/vercel/next.js/issues/40183

**Comment by Claudio-code** [comment]
I'm trying to apply the issue solution, but it's hard to find packages that use esm

**Comment by razzeee** [comment]
It also makes it sound like it's an regression on their end, so maybe if they fix it upstream we could wait for it.
I guess we should test if the downgrade like someone suggests in the issue helps in our case.

**Comment by Claudio-code** [comment]
I tried downgrading jest to 27 and next to 13.2 and 13.1 don't work, but I  if you remove node_modules and yarn.lock works but if you remove just remove node_modules not works,  I don't know why
